04-onion-routing.md

```diff
@@ -1407,6 +1438,117 @@ The _origin node_:
   - MAY use the data specified in the various failure types for debugging
   purposes.
 
+## Recommendations for Reputation
+
+Local reputation is used by forwarding nodes to decide whether to endorse a HTLC
+on their outgoing link. Nodes MAY use any metric of their choosing to classify
+a peer as having "good" reputation, though a poor choice of reputation scoring
+metric may affect their downstream reputation.
+
+Peers build reputation by forwarding successful HTLCs that resolve quickly, and
+lose reputation if they endorse failing or slow-resolving HTLCs. Reputation is
+only _negatively_ affected if an endorsed HTLC resolves undesirably, to hold
+nodes accountable for their endorsement signal while still allowing them to 
+forward unendorsed HTLCs that they are not certain about.
+
+HTLC resolution time is assessed relative to a threshold that the node 
+considers to be a reasonable amount of time for a HTLC to resolve: 
+- `resolution_period`: the amount of time a HTLC is allowed to resolve in that 
+  is classified as "good" behavior, expressed in seconds (default: 60 seconds).
+- `resolution_time`: the time elapsed between `update_add_htlc` and its 
+  resolution (`update_fulfill_hltc` / `update_fail_hltc` / 
+  `update_fail_malformed_htlc`), expressed in seconds.
+
+Each HTLC's contribution to reputation is expressed by its `effective_fees` 
+which capture the fees that it paid and the opportunity cost that holding it 
+for `resolution_time` incurred. 
+- `fees`: the fees paid by a forwarded HTLC (as described in [BOLT #7](07-routing-gossip.md#htlc-fees), 
+  equal to 0 if the HTLC was not fulfilled).
```

Comment:
I think I'm not understanding something. If `fees` are equal to zero for unfulfilled HTLCs, then it means that `opportunity_cost` is also zero. Does this mean failed HTLCs won't result in losing reputation?

Reply:
For zero fees a default `ppm` will be assumed (100ppm?) in order to bypass this.

Also, if we're talking about a **fast failing** HTLC then even if we have fees the loss is going to be zero (can be seen on `opportunity_cost` formula)

Reply:
Is that for when the payment is fulfilled but fees sent by the sender are zero? Or for the case mentioned here `equal to 0 if the HTLC was not fulfilled`?